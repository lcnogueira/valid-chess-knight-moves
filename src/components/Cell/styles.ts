import styled, { css } from 'styled-components'

type Props = {
  bgColor: string
}

export const Cell = styled.button<Props>`
  ${({ bgColor, theme: { colors } }) => css`
    background-color: ${bgColor};
    color: ${colors.white};
    border: none;
    text-decoration: none;
    height: 0;
    padding: 50% 0;
    display: flex;
    align-items: center;
    justify-content: center;
    border: none;

    :hover {
      opacity: 0.8;
    }
  `}
`
