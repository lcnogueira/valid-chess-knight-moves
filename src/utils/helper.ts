export const lines = ['8', '7', '6', '5', '4', '3', '2', '1']
export const columns = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H']

export const isValidNumber = (number: string) => lines.includes(number)

export const isValidLetter = (letter: string) =>
  columns.includes(letter.toUpperCase())

/**
 * The position is valid if componse by letter followed by one number and they both belong to the chess board limit.
 * @param position The position the user clicked on
 */
export const isAlgebraicNotation = (position: string) => {
  if (
    /^\D\d$/.test(position) &&
    isValidLetter(position.charAt(0)) &&
    isValidNumber(position.charAt(1))
  ) {
    return true
  }
  return false
}

/**
 * Returns the knight possible positions.
 * @param position The position the user clicked on
 */
export const findKnightPossibleMoves = (position: string): string[] => {
  if (position === 'A8') {
    return ['B6', 'C7']
  } else {
    return []
  }
}
