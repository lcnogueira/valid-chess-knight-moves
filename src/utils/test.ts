import {
  isAlgebraicNotation,
  isValidPosition,
  isValidIndex,
  findKnightPossibleMoves
} from './helper'

describe('helper methods', () => {
  test(`isAlgebraicNotation returns true if it's a valid notation`, () => {
    expect(isAlgebraicNotation('A1')).toBeTruthy()
    expect(isAlgebraicNotation('a1')).toBeTruthy()
    expect(isAlgebraicNotation('I1')).toBeTruthy()
    expect(isAlgebraicNotation('1A')).toBeFalsy()
    expect(isAlgebraicNotation('A11')).toBeFalsy()
  })

  test(`isValidPosition returns true if it's a valid notation and is in the chess board limit`, () => {
    expect(isValidPosition('A1')).toBeTruthy()
    expect(isValidPosition('A8')).toBeTruthy()
    expect(isValidPosition('h1')).toBeTruthy()
    expect(isValidPosition('H8')).toBeTruthy()
    expect(isValidPosition('A9')).toBeFalsy()
    expect(isValidPosition('H9')).toBeFalsy()
  })

  test('isValidIndex returns true if the index is equal or less than the array size', () => {
    expect(isValidIndex(0)).toBeTruthy()
    expect(isValidIndex(7)).toBeTruthy()
    expect(isValidIndex(-1)).toBeFalsy()
    expect(isValidIndex(8)).toBeFalsy()
  })

  test('findKnightPossibleMoves returns the posible positions where the knight can move', () => {
    expect(findKnightPossibleMoves('A8').sort()).toEqual(
      [
        'A4',
        'A6',
        'A8',
        'B5',
        'B6',
        'C4',
        'C7',
        'C8',
        'D5',
        'D7',
        'E6',
        'E8'
      ].sort()
    )
  })
})
