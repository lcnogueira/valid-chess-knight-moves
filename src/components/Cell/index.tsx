import * as S from './styles'
import { useColor } from 'hooks'

type Props = {
  isOdd: boolean
  onClick: (e: React.MouseEvent) => void
  isSelected: boolean
  isPossiblePosition: boolean
}

const Cell: React.FC<Props> = ({
  isOdd,
  onClick,
  isSelected = false,
  isPossiblePosition = false
}) => {
  const { colors } = useColor()
  let bgColor

  if (isSelected) {
    bgColor = colors.red
  } else if (isPossiblePosition) {
    bgColor = colors.green
  } else {
    bgColor = isOdd ? colors.secondary : colors.primary
  }

  return <S.Cell onClick={onClick} bgColor={bgColor} />
}

export default Cell
