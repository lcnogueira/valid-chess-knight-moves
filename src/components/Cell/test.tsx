import { render, screen, fireEvent } from 'tests/utils'

import Cell from '.'

const DefaultCell = ({
  isOdd = true,
  onClick = jest.fn(),
  isSelected = false,
  isPossiblePosition = false
}) => (
  <Cell
    isOdd={isOdd}
    onClick={onClick}
    isSelected={isSelected}
    isPossiblePosition={isPossiblePosition}
  >
    A8
  </Cell>
)

describe('<Cell />', () => {
  it('renders correctly with default props', () => {
    const { container } = render(<DefaultCell />)

    expect(container).toMatchSnapshot()
  })

  it('calls onClick prop when clicked', () => {
    const onClick = jest.fn()
    render(<DefaultCell onClick={onClick} />)

    fireEvent.click(screen.getByRole('button'))
    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
