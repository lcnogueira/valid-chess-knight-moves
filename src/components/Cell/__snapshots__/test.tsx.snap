// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cell /> renders correctly with default props 1`] = `
.c0 {
  background-color: var(--tomato);
  height: 0;
  padding-top: 100%;
  border: none;
}

.c0:hover {
  opacity: 0.8;
}

<div>
  <button
    class="c0"
  />
</div>
`
